### ISSUE TITLE
Model output is different when saved and loaded locally and in hf-hub

### ISSUE THREAD

**Issue by slice-pranay** [MaartenGr/BERTopic#1356]
Hi @MaartenGr ,
I am facing an issue where a model is giving different results in different scenarios with the same data. Here is the code which I used to get results in the scenarios mentioned below

```python
from bertopic import BERTopic
from cuml.cluster import HDBSCAN
from cuml.manifold import UMAP
import pandas as pd
from sklearn.datasets import fetch_20newsgroups
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

train = docs[:18000]
test = docs[18000:]

umap_model = UMAP(n_components=3, n_neighbors=10, min_dist=0.0, random_state=97)
hdbscan_model = HDBSCAN(min_samples=15, min_cluster_size=15, gen_min_span_tree=True, prediction_data = True)
topic_model = BERTopic(umap_model=umap_model, hdbscan_model=hdbscan_model, verbose=True)

t1,p1 = topic_model.fit_transform(train)

t11,p11 = topic_model.transform(train)

topic_model.save('models/test_model')
model2 = BERTopic.load('models/test_model')

t2,p2 = model2.transform(train)

topic_model.push_to_hf_hub('Pranay007/test_model', private=True, save_embedding_model=True, save_ctfidf=True)
model3 = BERTopic.load('Pranay007/test_model')

t3,p3 = model3.transform(train)
``` 

In this code I have trained a model with a public dataset. I have performed 
```
1. fit_transform()                                   ==> topics  are in variable t1, 
2. transform()                                       ==> topics  are in variable t11, 
3. save, load and transform()                        ==> topics  are in variable t2, 
4. saving to huggingface-hub, load and transform()   ==> topics  are in variable t3 
```
The observations are:

1. `t11` and `t2` are exactly same and `p11` and `p2` are also same so no issue here.
2. `t1` and `t11` have 40 mismatches. These 40 are assigned some topic in `fit_transform()` but they are assigned -1 in `transform()`. Still no issue for me but if solved it would be great.
3. `t1` and `t3` have almost 51% mismatches ( 9108 out of 18000 ). Almost all of the outliers in `t1` ( around 7K ) are assigned some clusters in `t3` and remaining have different clusters. Similar is the case with probabilities. `p1` and `p3` are all different and the difference in the probabilities is also huge for every document.

The versions of libraries i am using are 
bertopic==0.15.0
cuml-cu11==23.4.1
cudf-cu11==23.4.1
cuda toolkit 11.8

I am unable to understand why the huggingface-hub method is not performing like the normal saved model. Am I missing anything here. Please let me know how to make hf-hub model perform the same way as other model or if I am saving it wrong how to save it correctly.

Thanks in advance.

**Comment by MaartenGr** [comment]
There are a number of things happening here that are partly explained in the [documentation](https://maartengr.github.io/BERTopic/getting_started/serialization/serialization.html) but I'll go through a bit more in-depth here. 

To start off, saving your model like this:

```python
topic_model.save('models/test_model')
```
is quite different from saving your model like this:

```python
topic_model.push_to_hf_hub('Pranay007/test_model', private=True, save_embedding_model=True, save_ctfidf=True)
```

The main reason for this is that `.save` by default saves the model using `pickle` and `push_to_hub` saves it using pytorch. The main difference between those two is that pickle saves the full model, including HDBSCAN and UMAP, whereas pytorch skips over the embedding model, HDBSCAN, and UMAP since inference can be done with the topic embeddings instead. 

As a result, when you run `.transform` instead of having to go through HDBSCAN->UMAP, which can be costly, it applies inference on the embeddings directly instead by comparing cosine similarity. 

An advantage of using pytorch/safetensors is that the resulting model is at most 20Mb whereas using pickle can result in a couple of GBs.

**Comment by MaartenGr** [comment]
Closing this due to inactivity. Let me know if you want to re-open the issue!
